engine-mr/src/main/java/org/apache/kylin/engine/mr/steps/BuildGlobalHiveDictTotalBuildMapper.java

```diff
@@ -0,0 +1,231 @@
+        logger.info("global dic.{}.split.num.{} build dict start offset is {}", colName, partitionNum, start);
+    }
+
+    @Override
+    public void doMap(KEYIN key, Text record, Context context) throws IOException, InterruptedException {
+        long inkey = Long.parseLong(key.toString());
+        mos.write(colIndex+"", record, new LongWritable(start + inkey), "dict_column="+colName+"/"+colIndex);
+    }
+
+    @Override
+    protected void doCleanup(Context context) throws IOException, InterruptedException {
+        mos.close();
+    }
+
+    private Map<Integer, TreeMap<Integer, Long>> getPartitionsCount(Configuration conf, String partitionStatPath) throws IOException {
+        StringBuffer sb = new StringBuffer();
+        String temp=null;
+
+        String[] fileNameArr = null;
+        String[] statsArr = null;
+
+        //colStats key is last step reduce num,value is last step that reduce item count
+        TreeMap<Integer, Long> colStats = null;
+        //allStats key is colIndex,value is colStats(that col statistics info)
+        Map<Integer, TreeMap<Integer, Long>> allStats = new HashMap<>();
+
+        Path path = new Path(partitionStatPath);
+        FileSystem fs = path.getFileSystem(conf);
+        if (fs.exists(path) && fs.isDirectory(path)) {
+            for (FileStatus status : fs.listStatus(path)) {
+                //fileNameArr[0] is globaldict colIndex
+                fileNameArr=status.getPath().getName().split("-");
+                colStats = allStats.get(Integer.parseInt(fileNameArr[0]));
+                if(colStats==null){
+                    colStats = new TreeMap<>();
+                }
+                temp=cat(status.getPath(), fs);
+                logger.info("partitionStatPath:{},content:{}", partitionStatPath, temp);
+                if(temp!=null){
+                    statsArr = temp.split("\t");
+                    colStats.put(Integer.parseInt(statsArr[1]), Long.parseLong(statsArr[0]));
+                    allStats.put(Integer.parseInt(fileNameArr[0]), colStats);
+                }
+            }
+        }
+
+        allStats.forEach((k, v)->{
+            v.forEach((k1, v1)->{
+                logger.info("allStats.colIndex:{},this split num:{},this split num's count:{}", k, k1, v1);
+            });
+        });
+
+        return allStats;
+    }
+
+    private String cat(Path remotePath, FileSystem fs) throws IOException {
+        FSDataInputStream in = null;
+        BufferedReader buffer = null;
+        StringBuffer stat= new StringBuffer();
+        try {
+            in= fs.open(remotePath);
+            buffer= new BufferedReader(new InputStreamReader(in, "UTF-8")) ;
+            String line = null;
+            while ((line = buffer.readLine()) != null) {
+                stat.append(line);
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }finally {
+            buffer.close();
+            in.close();
```

Comment:
Maybe checkIfNotNull before close it for `in` and `buffer` is better, see https://lgtm.com/projects/g/apache/kylin/rev/pr-41895d7e0679c810b18a84e71eb8221f677676e0.

Reply:
OK，thanks.